okio/src/main/java/okio/Buffer.java

```diff
@@ -793,6 +869,42 @@ public void clear() {
     return this;
   }
 
+  @Override public Buffer writeUtf8CodePoint(int codePoint) {
+    if (codePoint < 0x80) {
+      // Emit a 7-bit code point with 1 byte.
+      writeByte(codePoint);
+
+    } else if (codePoint < 0x800) {
+      // Emit a 11-bit code point with 2 bytes.
+      writeByte(codePoint >>  6        | 0xc0); // 110xxxxx
+      writeByte(codePoint       & 0x3f | 0x80); // 10xxxxxx
+
+    } else if (codePoint < 0x10000) {
+      if (codePoint >= 0xd800 && codePoint <= 0xdfff) {
+        throw new IllegalArgumentException(
```

Comment:
Interesting design decision here. What to do when the input's not quite right:
 - we could throw
 - we could encode it (it's still recoverable on the other end, though encoders might return the replacement char)
 - we could encode the replacement character ourselves

Thoughts?